File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
          # use hash map key: node label value: new node
        # time complexity o(V + E)
        if not node:
            return None

        labels_to_nodes = {}
        stack = [node]
        visited = set([node.val])

        while stack:
            n = stack.pop()

            if n.val not in labels_to_nodes:
                labels_to_nodes[n.val] = Node(val=n.val)

            for nbr in n.neighbors:
                if nbr.val not in labels_to_nodes:
                    labels_to_nodes[nbr.val] = Node(val=nbr.val)
                labels_to_nodes[n.val].neighbors.append(labels_to_nodes[nbr.val])
                
                if nbr.val not in visited:
                    stack.append(nbr)
                    visited.add(nbr.val)

        return labels_to_nodes[1]

        
        

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) {
            return nullptr;
        }
        unordered_map<int, Node*> copiedNodes;

        unordered_set<int> visited;
        stack<Node*> nodes;

        nodes.push(node);

        while (!nodes.empty()) {
            auto *node = nodes.top();
            nodes.pop();
            if (visited.find(node->val) != visited.end()) {
                continue;
            }

            visited.insert(node->val);

            if (copiedNodes.find(node->val) == copiedNodes.end()) {
                copiedNodes[node->val] = new Node(node->val);
            }

            for (auto &neighbor: node->neighbors) {
                if (visited.find(neighbor->val) == visited.end()) {
                    nodes.push(neighbor);
                }
                if (copiedNodes.find(neighbor->val) == copiedNodes.end()) {
                    copiedNodes[neighbor->val] = new Node(neighbor->val);
                }
                copiedNodes[node->val]->neighbors.push_back(copiedNodes[neighbor->val]);
            }            
                    
        }
        return copiedNodes[node->val];
    } 
};
